### ISSUE TITLE
Output file option

### ISSUE THREAD

**Issue by tw3lveparsecs** [Azure/arm-ttk#110]
It would be nice if we had the option to output the pester results to file and choose our output format, e.g. `Test-AzTemplate -TemplatePath $TemplateFileOrFolder -OutputFormat  NUnitXml -OutputFile .\results.xml` 

Will be especially helpful when publishing test results to Azure DevOps. 
I know there is a extension that can be installed for Azure DevOps and other workarounds available but a native solution would be awesome.

**Comment by MartinAngelov** [comment]
@tw3lveparsecs . If you don't want or can't use extensions in your AzDO organization, you can create a new `*.test.ps1` file that does all the preparation (e.g. importing arm-ttk module) and calls `Test-AzTemplate `cmd-let with `-Pester `switch and then use:

`Invoke-Pester -Script <YourScriptWithTestAzTemplate>.test.ps1 -OutputFormat NUnitXml -OutputFile <NUnitXmlFilePath>`  

Wrap this in a PowerShell task into your AzDO and use another one to upload the results.

**Comment by SacredSkull** [comment]
@MartinAngelov 
I'm performing the above, with this script:
```
Import-Module -Force -Global '.../arm-ttk.psd1'
Test-AzTemplate -TemplatePath ".../ARM" -Pester
```

(the ... means a long chain of directories, these folders do exist)

But when pester actually tries to run these tests:
```
 CommandNotFoundException: The term 'Test-Case' is not recognized as the name of a cmdlet, function, script file, or operable program.
 Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

I tried creating an empty `function Test-Case {}` in the same scope as the `Test-AzTemplate` commandlet and it stops complaining anymore (as you'd expect).
I'm not sure you can actually do this because `Test-Case` is an inner function for ARM-TTK, nested inside the main/exported `Test-AzTemplate` - it seems that the extension must be doing something else to get around this, because as is, this isn't actually possible (AFAIK).

**Comment by tw3lveparsecs** [comment]
@MartinAngelov @bmoore-msft 
I am happy to put this one down to my own ignorance but i can seem to get this working.

I have a simple ARM template with no params as per below.

```json
{
    "$schema": "https://schema.management.azure.com/schemas/2018-05-01/subscriptionDeploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "resources": [
        {
            "type": "Microsoft.Insights/diagnosticSettings",
            "apiVersion": "2017-05-01-preview",
            "name": "ActivityLogs",
            "location": "global"
        }
    ]
}
```
When i run command below it returns a fail as expected.

```powershell
Test-AzTemplate -TemplatePath .\ARMTemplate.json -Test "Parameters Property Must Exist"

Validating Temp\ARMTemplate.json
  deploymentTemplate
    [-] Parameters Property Must Exist (6 ms)
        Parameters property must exist in the template`
```
When I run the same command with the -Pester switch, it returns the following. Am I doing something wrong?
```powershell
Test-AzTemplate -TemplatePath .\ARMTemplate.json -Test "Parameters Property Must Exist" -Pester

Starting discovery in 1 files.
Discovery finished in 10ms.
Tests completed in 8ms
Tests Passed: 0, Failed: 0, Skipped: 0 NotRun: 0
```

**Comment by MartinAngelov** [comment]
@SacredSkull , What version of Pester module do you use ? Version 5 wont work at the moment (and it is still in alpha).

Try forcing using of pester 4.10.2 (also tested with lower) .

**Comment by MartinAngelov** [comment]
> @MartinAngelov @bmoore-msft
> I am happy to put this one down to my own ignorance but i can seem to get this working.
> 
> I have a simple ARM template with no params as per below.
> 
> ```json
> {
>     "$schema": "https://schema.management.azure.com/schemas/2018-05-01/subscriptionDeploymentTemplate.json#",
>     "contentVersion": "1.0.0.0",
>     "resources": [
>         {
>             "type": "Microsoft.Insights/diagnosticSettings",
>             "apiVersion": "2017-05-01-preview",
>             "name": "ActivityLogs",
>             "location": "global"
>         }
>     ]
> }
> ```
> 
> When i run command below it returns a fail as expected.
> 
> ```powershell
> Test-AzTemplate -TemplatePath .\ARMTemplate.json -Test "Parameters Property Must Exist"
> 
> Validating Temp\ARMTemplate.json
>   deploymentTemplate
>     [-] Parameters Property Must Exist (6 ms)
>         Parameters property must exist in the template`
> ```
> 
> When I run the same command with the -Pester switch, it returns the following. Am I doing something wrong?
> 
> ```powershell
> Test-AzTemplate -TemplatePath .\ARMTemplate.json -Test "Parameters Property Must Exist" -Pester
> 
> Starting discovery in 1 files.
> Discovery finished in 10ms.
> Tests completed in 8ms
> Tests Passed: 0, Failed: 0, Skipped: 0 NotRun: 0
> ```

Is this related to the initial prompt or it is separate issue on its own ?

**Comment by tw3lveparsecs** [comment]
@MartinAngelov was getting issue due to running pester version 5, when running pester version 4.10.1 it works. Thanks!

**Comment by tw3lveparsecs** [comment]
@MartinAngelov just to close the loop when using pester version 4, i have been able to get this working in Azure DevOps natively using the method you suggested previously.

**Comment by SacredSkull** [comment]
Yep forcing Pester to be version 4.10.1 works (you mentioned 4.10.2 which doesn't seem to exist on PSGallery) - I created this little snippet to ensure that the imported version is correct if this helps anyone:

```
try {
    Remove-Module -Force -Name Pester -ErrorAction SilentlyContinue
    Import-Module -Force -Name Pester -RequiredVersion 4.10.1
} catch {
    Install-Module -Name Pester -Force -RequiredVersion 4.10.1
    Import-Module -Name Pester -RequiredVersion 4.10.1
}
```

**Comment by asears** [comment]
Is this fixed for pester 5 now?  If I run 2 consecutive Test-AzTemplate commands with -Pester as option, it seems to run the first fine, however second test fails with following error:

```powershell
Add-Type: C:\Program Files\WindowsPowerShell\Modules\Pester\5.0.4\Pester.psm1:109
Line |
 109 |      Add-Type -Path "$PSScriptRoot/bin/netstandard2.0/Pester.dll"
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Assembly with same name is already loaded
```

Should check for loaded Pester prior to reimport.

**Comment by StartAutomating** [comment]
@bmoore-msft Providing the provided output parameters wouldn't be that bad.  Trying to make things work in Pester 5 is likely to be trickier, and should be a separate issue.
